### ISSUE TITLE
is it possible to estimate a non-binary cjoint outcome?

### ISSUE THREAD

**Issue by SimonMontfort** [tsrobinson/cjbart#3]
Is it possible to estimate a cjbart for a conjoint outcome for non-binary outcomes, e.g. continuous/likert scale?

Data setup:

     library(cjbart)
     subjects <- 250
     rounds <- 5
     profiles <- 2
     obs <- subjects*rounds*profiles

     fake_data <- data.frame(A = sample(c("a1","a2","a3"), obs, replace = TRUE),
                             B = sample(c("b1","b2","b3"), obs, replace = TRUE),
                             C = sample(c("c1","c2","c3"), obs, replace = TRUE),
                             D = sample(c("d1","d2","d3"), obs, replace = TRUE),
                             E = sample(c("e1","e2","e3"), obs, replace = TRUE),
                        
                             covar1 = rep(runif(subjects, 0 ,1),
                                          each = rounds),
                             covar2 = rep(sample(c(1,0),
                                                 subjects,
                                                 replace = TRUE),
                                          each = rounds),
                        
                             id1 = rep(1:subjects, each=rounds),
                             stringsAsFactors = TRUE)

in the example in `?cjbart` this could be an outcome defined in the following way:

     fake_data$Y <- ifelse(fake_data$E == "e2",
                           rbinom(obs, 1, fake_data$covar1),
                           sample(c(1,2,3,4,5), obs, replace = TRUE))

Note that the outcome is an integer between 1 and 5 instead of a binary outcome of 0 or 1.

using this different outcome provides the following error when running:

     cj_model <- cjbart::cjbart(data = fake_data,
                                Y = "Y", 
                                id = "id1")
     het_effects <- cjbart::IMCE(data = fake_data, 
                                 model = cj_model, 
                                 attribs = c("A","B","C","D","E"),
                                 ref_levels = c("a1","b1","c1","d1","e1"),
                                 cores = 1)

     Calculating OMCEs for attribute: A [1/5]
     Error in quantile.default(var_z[, data[[id]] == s], intvl) : 
       missing values and NaN's not allowed if 'na.rm' is FALSE

I've noticed that if I rescale the five levels to values between 0 and 1, the algorithm works, e.g. using `c(0, 0.25, 0.5, 0.75,  1)` when sampling to generate `Y` the algorithm works. What does this imply for the estimation? Is this a valid approach for use with non-binary outcomes? It seems that the package was not designed for non-binary outcomes.

**Comment by tsrobinson** [comment]
Great suggestion @SimonMontfort!  We can certainly implement this functionality in the next release.

The reason it errors currently is because we are using a probit version of BART (BART::mc.pbart), so the outcome must be within the 0-1 range.

We can, however, add an extra argument to the cjbart function that toggles between this probit version and a continuous version (BART::mc.wbart).

**Comment by tsrobinson** [comment]
Steps to implement:

- [ ] Add `outcome` argument to cjbart()` with three options: `NULL` (default; detect outcome range automatically), `"choice"`, and `"rating"`
- [ ] Update IMCE() function to extract correct vector from prediction objects
- [ ] Update AMCE() 
- [ ] Update pIMCE()

**Comment by tsrobinson** [comment]
@SimonMontfort -- this should now be implemented (though currently untested). If you wanted to give it an early try, you can install the development branch for this issue by running:

`devtools::install_github("https://github.com/tsrobinson/cjbart/tree/wbart")`

**Comment by SimonMontfort** [comment]
Thank you @tsrobinson! That's really great.

I've tried with the example above but get the following error:

    Calculating OMCEs for attribute: A [1/5]
    Error: $ operator is invalid for atomic vectors

I've installed from github as you suggested.

**Comment by tsrobinson** [comment]
Thanks @SimonMontfort  -- fixed the issue. 

The following MWE runs both a binary and continuous version of the method, and should replicate if you reinstall cjbart using the above devtools code:

```
subjects <- 250
rounds <- 5
profiles <- 2
obs <- subjects*rounds*profiles

fake_data <- data.frame(A = sample(c("a1","a2","a3"), obs, replace = TRUE),
                        B = sample(c("b1","b2","b3"), obs, replace = TRUE),
                        C = sample(c("c1","c2","c3"), obs, replace = TRUE),
                        D = sample(c("d1","d2","d3"), obs, replace = TRUE),
                        E = sample(c("e1","e2","e3"), obs, replace = TRUE),
                        
                        covar1 = rep(runif(subjects, 0 ,1),
                                     each = rounds),
                        covar2 = rep(sample(c(1,0),
                                            subjects,
                                            replace = TRUE),
                                     each = rounds),
                        
                        id1 = rep(1:subjects, each=rounds),
                        stringsAsFactors = TRUE)

fake_data$Y_c <- sample(c(1,2,3,4,5), nrow(fake_data), replace = TRUE)
fake_data$Y_b <- sample(c(0,1), nrow(fake_data), replace = TRUE)

data_c <- fake_data[,names(fake_data) != "Y_b"]
data_b <- fake_data[,names(fake_data) != "Y_c"]

mod_c <- cjbart::cjbart(data = data_c,
                        Y = "Y_c", 
                        id = "id1")

mod_b <- cjbart::cjbart(data = data_b,
                        Y = "Y_b", 
                        id = "id1")

het_c <- cjbart::IMCE(data = data_c, 
                      model = mod_c, 
                      attribs = c("A","B","C","D","E"),
                      ref_levels = c("a1","b1","c1","d1","e1"),
                      cores = 8)

het_b <- cjbart::IMCE(data = data_b, 
                      model = mod_b, 
                      attribs = c("A","B","C","D","E"),
                      ref_levels = c("a1","b1","c1","d1","e1"),
                      cores = 8)


```

**Comment by tsrobinson** [comment]
Resolved and pushed to CRAN under v0.3.0
